handler/src/main/java/io/netty/handler/ssl/SslHandler.java

```diff
@@ -2027,6 +2049,9 @@ private void setHandshakeFailureTransportFailure(ChannelHandlerContext ctx, Thro
     }
 
     private void releaseAndFailAll(ChannelHandlerContext ctx, Throwable cause) {
+        if (resumptionController != null) {
+            resumptionController.remove(engine());
```

Comment:
hmm... is this really what we want ? I think we only want to remove it if `isValid()` is false or if the exception is a `SslHandshakeException`

Reply:
Calls to this method are always associated with failing the handshake promise. The promise can be reinitialized by a renegotiation, though I'm not sure if that's a possibility after we get here.
If we get here from `setHandshakeFailureTransportFailure` then the exception won't be an `SslHandshakeException` for instance.

If we filter out too many remove calls, then the failure mode is that an engine will remain in the `confirmedValidations` and can prevent a `resume` call from happening on session resumption. If we filter out too few remove calls, then if somehow a `remove` call happened before a `validateResumeIfNeeded` call, then we'd call both a `check*` method and a `resume*` method on the trust manager.

The latter seems to be a possibility e.g. in this stretch of code https://github.com/netty/netty/blob/4.1/handler/src/main/java/io/netty/handler/ssl/SslHandler.java#L973-L997 because the remove is not guarded by changing the handshake promise state. Perhaps that's the real fix? That the `remove` call should follow `handshakePromise.tryFailure`?